carapace-server/src/main/java/org/carapaceproxy/server/mapper/StandardEndpointMapper.java

```diff
@@ -367,37 +377,39 @@ public MapResult map(HttpRequest request, String userId, String sessionId, Backe
                     selectedBackends = backendSelector.selectBackends(userId, sessionId, director, matchResult);
                 }
                 somethingMatched = somethingMatched | !selectedBackends.isEmpty();
+
                 LOG.log(Level.FINEST, "selected {0} backends for {1}, director is {2}", new Object[]{selectedBackends, request.uri(), director});
                 for (String backendId : selectedBackends) {
+                    Action selectedAction;
                     switch (action.getType()) {
-                        case ActionConfiguration.TYPE_PROXY: {
-                            BackendConfiguration backend = this.backends.get(backendId);
-                            if (backend != null && backendHealthManager.isAvailable(backendId)) {
-                                return new MapResult(
-                                        backend.getHost(),
-                                        backend.getPort(),
-                                        MapResult.Action.PROXY,
-                                        route.getId(),
-                                        action.getCustomHeaders()
-                                );
-                            }
+                        case ActionConfiguration.TYPE_PROXY:
+                            selectedAction = MapResult.Action.PROXY;
                             break;
-                        }
                         case ActionConfiguration.TYPE_CACHE:
-                            BackendConfiguration backend = this.backends.get(backendId);
-                            if (backend != null && backendHealthManager.isAvailable(backendId)) {
-                                return new MapResult(
-                                        backend.getHost(),
-                                        backend.getPort(),
-                                        MapResult.Action.CACHE,
-                                        route.getId(),
-                                        action.getCustomHeaders()
-                                );
-                            }
+                            selectedAction = MapResult.Action.CACHE;
                             break;
                         default:
                             return MapResult.NOT_FOUND(route.getId());
                     }
+
+                    BackendConfiguration backend = this.backends.get(backendId);
+                    if (backend != null && backendHealthManager.isAvailable(backendId)) {
+                        List<CustomHeader> customHeaders = new ArrayList(action.getCustomHeaders());
+                        if (this.debuggingHeaderEnabled) {
```

Comment:
if this debug is not enabled you can skip the clone of the arraylist.

please also ensure that action.getCustomHeaders() returns an unmodifiableList